torch/sparse/semi_structured.py

```diff
@@ -0,0 +1,318 @@
+from collections import namedtuple
+import warnings
+
+import torch
+
+
+__all__ = [
+    "to_sparse_semi_structured",
+    "SparseSemiStructuredTensor",
+]
+
+CUSPARSELT_CONFIG = namedtuple("CUSPARELT_CONFIG", "compression_factor n min_size")
+DTYPE_TO_CUSPARSELT_CONFIG = {
+    torch.float32 : CUSPARSELT_CONFIG(9, 1, 8),
+    torch.float16 : CUSPARSELT_CONFIG(9, 2, 16),
+    torch.bfloat16 : CUSPARSELT_CONFIG(9, 2, 16),
+    torch.int8: CUSPARSELT_CONFIG(10, 4, 32),
+}
+
+class SparseSemiStructuredTensor(torch.Tensor):
+    """
+    This class implementes semi-structured (2:4) sparsity as a Tensor subclass.
+
+    Semi-structured sparsity describes a sparsity pattern where n in every 2n elements are sparse,
+    depending on the datatype. It is most commonly referred to as 2:4 sparsity or fine-grained
+    structured sparsity.
+
+    compressed_tensor is a tensor that stores both the kept elemenets and the metadata mask
+    These two are stored next to each other in one contiguous tensor.
+
+    compressed tensor = [ kept elemetns of original tensor |   mask_metadata     ]
+
+    For an original tensor of size (m, k) we expect the first m * k // 3 elements to be the kept elements
+    The rest of the tensor is metadata.
+
+    This tensor subclass also has a cslt object, which is set if pytorch is built with cuSPARSELt support.
+    In this case, we store some additional metadata containg the sparse matrix descriptor in order for
+    faster matmul performance.
+
+    In the future, when this cslt object is not set, we can use _structured_sparse_linear.
```

Comment:
Not sure what this means?
Also generally: public API docs should only refer to public APIs.